### ISSUE TITLE
Extended weapons

### ISSUE THREAD

**Issue by LemADEC** [LemADEC/WarpDrive#36]
We're missing alternate ways to defend or attack with ships. This issue is tracking possible ideas to address that, in no special order...
General
- Add a friend or foe system for automated weapons

Gun turret
- Add a cheap and easy to setup defense laser

=> ICBM and open turrets can do that

Drones
- Add a drone controller block, each block can controller 1 drone
  - has no power requirement or has its own battery to counter for EMP/lasers
  - onBreaking, drone is looking for a new friendly 'free' controller or start loosing health
- Add a drone entity capable of shooting other enemies, returning to controller
  - if controller is too far, try to reach it and loose health
- Add a drone item consumed to spawn a new drone
- Need to limit the number of drones => one per player registered on the ship? slots system?
- Control station allows to select target: power (blue), movement (green) or weapon (red) system
- Possible leveling: specialization (size/power vs slots/mobility)

Foam bomb
- Add a laser option/missile to deploy foam on the enemy ship
- Foam is opaque, disabling the solar panels
- Foam decay after 180 seconds
- Possible leveling: consume hull, disable optics (camera/lasers), block new drone spawning

=> use gascraft or glenn's gases framework?

Entity repulsor
- Add a powered block to repulse players (instead of killing them like interdiction matrix)
- Block is incompatible with drones on the same ship (obviously)
- A 'key' item allows to enter the vicinity

=> already available in MFFS and ARS through damages, possibly extend those mods instead?

EMP coils
- Add a charging block that releasing an EMP disabling nearby ships, including emitter
- It's a counter for a Warp field inhibitor and lasers

EMP shield
- Counter the EMP shield but also prevents on-board Warp inhibitor, so you can still use lasers?

**Comment by TheNumenorean** [comment]
General: Fine
Gun turret: Sure, seems reasonable
Drones: This seems really really out of context here. I would say that that make much more sense for a different mod entirely.
Foam Bomb: I don't like the idea, breaks too many other mechanics, and will be really laggy. Essentially makes ICBM worthless.
Repulsor: No feelings realy
Emp: already done by ICBM

I am soon going to start updating ICBM to 1.7 as well, so I don't think we really need to add much more weapons. But maybe a proton torpedo? And a phaser? (I already have phaser code in a small personal mod, can just move that in)

**Comment by LemADEC** [comment]
How would you define a Proton torpedo and a Phaser vs the existing Laser cannon, for example?

All those are to give a more strategic/PvP dimension to the mod. They give more options and counters. As for the 'slots' idea, it's to encourage creativity, avoiding the 'generic does everything' ship.

**Comment by TheNumenorean** [comment]
Proton torpedo: fire and forget horizontal unguided missile
Phaser: Hand held laser

I still don't understand, slots in what?

**Comment by TheNumenorean** [comment]
Photon torpedo is more of a shield destroyer, which are currently invulnerable to lasers

**Comment by LemADEC** [comment]
Slots have no physical representation in the minecraft world. They're enabler for your ship.

I did a bit of research, here's a few additions:
- Magnetosphere = create a magnetic field around the ship, allows to move closer to sun, no ballistic protection
  - IonicGasDiffuser = project ionic gas in the magnetosphere (air protects from microwaves, Water protects from laser)
  - GasIoniser = produces Ionic gas
- Reflective hull = an hull with silver plating that absorbs 90% of laser energy
- PhaserCannon = shoots microwaves, more energy efficient than laser, slower to fire
  - PhaserBank = produces the energy, requires toxic gas, breaking it release the gas
  - PhaserLense = focalize the phaser beam to increase its power output
  - PhaserBankCooler = increase cool rate with a small added power consumption
  - bank blocks forms a parallelogram shaped structure
- DisruptorCannon = same as phaser, more efficient on shield than hull
- KineticCannon = good against hull and magnetoshield
- SpatialTordedoLauncher
  - SpatialTorpedo = fly-by-wire kinetic load
- PhotonTordepo = antimatter load with a shielded hull
- PlasmaTorpedo = short range, wide effect area, good vs shields
- PhasedTorpedos = torpedos able to pass through force field, unstable, slow movement

**Comment by lellasone** [comment]
Hello, 
I am one of the founding moderators for a small military-industrial server that runs with an emphasis on ship to ship combat (we started during the days of repower frames and have since adopted RIM and WD). Since WD combat is something that matters a lot to us I wanted to chime in with some thoughts from myself and others on the server. 

General Thoughts: 
The big thing for me as a ship designer would be that in-game interactions almost universally make for richer more interesting mechanics than interactions within mods.  What you are talking about with slots sounds interesting, but that kind of magic block system is almost never as interesting as building up integration to create emergent mechanics and seems like it would take away a lot of the magic of building up ships that work in-world. 

Feedback on concepts: 
General – an friend/foe system would be very useful, particularly if it played nice with CC/OC.

Gun Turret – This sounds very exciting, particularly if they featured armor piercing damage they could 
function as both anti-entity and anti-player defenses and would present a much more dynamic experience than simple kill fields or anti-access fields.

Drones – player time is always at a premium so maybe not the most feasible option there, but potentially pretty useful if they were in some way warp equipped (for coordinating missile volleys and the like)

Foam Bomb – Unless this had a very small area of effect it seems like this would encourage stalemates over combat (which is usually sub-optimal). It also seems like this would be more appropriate to ICBM which already has a robust system for handling missiles. 

Entity Repulsor – This has been done a number of other places, the role is a good one though. Perhaps turrets could serve to create a similar effect (with a bit more play in terms of allowing players to overwhelm defenses rather than just a hard limit). 

EMP Coils – it really feels like this could be done far better by having ICBM integration as that is already a robust system (and includes missiles). Having a mod specific EMP is kind of a pain since it forces players into really narrow “correct” ship builds. 

EMP Shield – although a cool concept, it seems like this would take away a lot of interesting 
gameplay around having to build hardened systems and work around the effects of an EMP. A different route might be to selectively make some of WDs systems vulnerable (have drives and cloaking be fully vulnerable but have lasers take a 15% hit for example). 

Below are some thoughts from myself and others on the server about what would help to make the combat experience more fleshed out and engaging:  

Warp: 
• Making jump costs slightly exponential with distance so that a 2k jump might cost say 6 times as much as a 1k jump could make for more interesting design choices in building chase ships. 

• Tying ship size to per-block jump costs would give smaller ships a bigger mobility edge and help enable tactical warping for smaller ships. 

Warp Inhibition: 
• Allowing players to prevent missiles from jumping inside their ship would help balance out tactical warp missiles. (perhaps there could be an EU cost equal to the cost of the jump blocked or some such)

• Making warp cores vulnerable to ICBM EMP’s would give players a robust and complex set of mechanics around trapping ships without making escape wholly impossible. 

• Another direction would be to have a warp inhibitor that required setup (say 4 corners type thing) to make it viable against larger ships and bases but keep it from being the death knell to frigate/cruiser sized vehicles. 

Lasers: 
• Adding mechanics where larger lasers spend more of their energy in explosion (vs penetration) could make for more interesting design and combat choices about how large of a laser to use. (It would also make all those variable length designs much more justifiable)

• Similarly adding some kind of block vs shield tradeoff with frequency could make for much more interesting tactical combat. (having shield blocks treated as say 15 blocks of reinforced stone with that number variable with frequency at a trade off against block damage efficiency)

Redstone:
• Probably the biggest thing that could be done to better enable combat would be to have blocks like the laser and warp core interface with Redstone. Being able to trigger a laser firing or jump from Redstone would make those blocks a lot more accessible to people who are not necessarily familiar with CC (it would also make setups a bit less lagy but that is a lessor issue)

TLDR: duplicating other mods tends to damage gameplay options, shield piercing lasers would be good, exponential jump costs with distance would make things more interesting, integration between mods always encourages growth.

**Comment by LemADEC** [comment]
Thanks for your valuable comments!
What's your modpack?

I'm talking about 'drone weapon', you're thinking about 'automated drone ship'. You can already do the latest with LUA scripts. It's a bit tedious, but doable.

I'm not happy on current proposal of Foam/goo/umbrella weapon. Still looking for something simpler to implement a solar panel counter.

Proposed EMP coils are multiblock. The ICBM turret and explosives are too easy to use and don't integrate well with other mods (many energy storage item/blocks remain untouched by it).

Regarding warp cost, it's currently = 'ship volume in block' x 'cost per block' + 'jump distance' x 'cost per m'. Hence, I don't quite get your point on that. I've created #49 for this subject.

Warp inhibitor is discussed in #17.

Laser cannon effects can be discussed in #48.

Yes, redstone is missing, hence #50 has been created.

**Comment by LemADEC** [comment]
More thoughts...

Gun turret:
A cheap version of the drones, with limited direction (?), slower fire and lower damage.
Weapon AI cores are required to control them.

Drones:
They're to defend from players and missiles, not from laser.
The would only work in space/hyperspace.
Dead drone could drop as an item.
Add recall button to use before jump, where they remains 3 blocks from their controller. Need to grab them during jump.

Slots:
Could be a multiple blocks 'Mainframe'.
We need a weapon controller to choose which system to enable/disable, etc.
All blocks are linked to ship core, so they link to each other.
Weapon mainframe automatically reload weapons and select targets, enables drone defense system.

Forcefield:
Use additive projectors so we can do localized damages.
Each AI core enables x? amount forcefield projector, hence avoiding the hull made of only projectors.
We can't limit the distance between projectors, that would be obscure to players.
Detecting the forcefield shape would be floodfill coating with a blacklist non-hull blocks to detect the inside (hull breach vs X-wing or halo shape).
Blacklist would include cables/conduits, torches, etc.

EMP coil:
Effect should be sustained for 10 to 20s. It needs to cover other mods, notably armors and jetpacks.
It should also affect other EMP coils, so you can't "spam" them.

EMP shield:
Would be a reactor and AI upgrades to protect them from EMP. No protection for warp core or other blocks.

Magnetosphere:
Basically a cheap reconfigurable kinetic shield. Might just changing frequency of it. Maybe use it as ship upgrade to go into liquids (collect deuterium from a star). Works with gun turrets, incompatible with laser, force field and drones. Good for mining ships, cheaper than forcefield, weak for attack ships.

Reflective hull:
Block made from normal hull block + glass (cheap). Can be broken with kinetic/explosive weapons, decaying to its normal/non-reflective state.
It's a cheap defense compared to magnetosphere.

**Comment by lellasone** [comment]
You mention hull blocks, are you thinking of having a dedicated block for ships to be made of?

**Comment by LemADEC** [comment]
Others mods only provide limited option in terms of hull decoration, we'll need eventually to provides ours. There's no intention to force players to use our blocks, just need to give them more options than tile entities (Carpenter blocks are bad for FPS) and IC2 Reinforced stone / ICBM Blast proof gray block.

**Comment by TheNumenorean** [comment]
I like the idea of having more to build with. It would also be nice to not just have to have a 5 re stone thick hull. Some sort of field-plating? Stronger if energized?

**Comment by lellasone** [comment]
Having more build options would be a very good thing. It would be nice if a bit of the current cost balance were preserved though, since It is kind of nice to have the N thick RE stone hull option as a viable way to go for pure bulk resistance.

Reflective Hull: Sounds like it would be rather very awesome as an option.

Energized field plating also sounds like it would be very shiny. Might be interesting to explore the possibility of having resistance grow by a root or log function (to make for those wonderful "increase power to defenses" moments)

**Comment by CaptainMegavolt** [comment]
I know there are already force field mods out there, but I thought that perhaps WarpDrive should have its own shield system, in which a shield's area is defined much like the dimensions of a ship. Shields are invisible unless hit, in which case they appear in that area. Shields require shield projectors (which emit the shield; multiple projectors can be configured to create larger and/or interestingly shaped shields), shield capacitors (which store energy for the shield), and a shield core (which this the block CC or OC computers would interact with to control the shield; the core needs to be linked to all shield projectors and capacitors with some kind of conduit. Note that contiguous capacitor clusters only need to be connected to with one conduit, rather like ME Drives from Applied Energistcs.) Every time a shield is hit, the capacitors lose some energy, so the more capacitors you have, the better. Some phasing weapons can penetrate shields (like suggested a few times in this thread), and EMPs can disable them temporarily. Electricity would be fed into the shield core block.

**Comment by CaptainMegavolt** [comment]
On the subject of weapons, it would be interesting to have a variety of different sorts of weapon systems. Personally, I was thinking of a few different types of weapon systems. One idea was the Missile Pod. This is a turret type thing that can launch a variety of missiles. Missiles have some modularity, so that you can mix and match engines, warheads, and guidance systems. (So for example you could have a missile with a low-acceleration high-efficiency ion engine, an advanced guidance system with long range tracking capability, and a radar jammer module. This sort of missile would be for chasing down ships at long range without them noticing, then completely destroying them with an antimatter explosion.)  Raised missiles, there could also be other sorts of turrets. Projectile weapons ( Like Space Engineers Gatling turrets), plasma cannons (shoot a bolt of hydrogen plasma), particle cannons (internal particle accelerator shoots out a beam of protons), and things like that. Also, drones. Different sorts of dro es for different jobs, such as sabotage drones that burrow through an enemy ships hull and try to destroy machinery, fighter drones which are fast, maneuverable, and equipped with some sort of light weapons, and defense drones which are slower, heavily armored, have shields, and have some sort of powerful main weapon. 
Besides obvious weapons, a few other weapon-type things could be added. Tractor beams for one (these seem to be a popular suggestion). Perhaps also phase-shift devices that enable ships to warp through solid blocks (but obviously not stop inside blocks). Another possibility would be radar jammers that prevent all radars in their effective area from functioning (although they can be set up to have a sort of safe zone so that the radar of the ship they're on still works. Last but not least, warp interdictors. Instead of just prevention g ships in their effective area from warping, they also knock ships that attempt to pass through their area out of warp, Elite Dangerous style. Something like this could exist for Hyperspace e interdiction too.
